internal/bridge/crunchybridgecluster/postgres.go

```diff
@@ -0,0 +1,174 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crunchydata/postgres-operator/internal/bridge"
+	"github.com/crunchydata/postgres-operator/internal/initialize"
+	"github.com/crunchydata/postgres-operator/internal/naming"
+	"github.com/crunchydata/postgres-operator/pkg/apis/postgres-operator.crunchydata.com/v1beta1"
+)
+
+// generatePostgresRoleSecret returns a Secret containing a password and
+// connection details for the appropriate database.
+func (r *CrunchyBridgeClusterReconciler) generatePostgresRoleSecret(
+	cluster *v1beta1.CrunchyBridgeCluster, roleSpec *v1beta1.CrunchyBridgeClusterRoleSpec,
+	clusterRole *bridge.ClusterRoleApiResource,
+) (*corev1.Secret, error) {
+	roleName := roleSpec.Name
+	secretName := roleSpec.SecretName
+	intent := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+		Namespace: cluster.Namespace,
+		Name:      secretName,
+	}}
+	intent.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Secret"))
+	initialize.StringMap(&intent.StringData)
+
+	intent.StringData["name"] = clusterRole.Name
+	intent.StringData["password"] = clusterRole.Password
+	intent.StringData["uri"] = clusterRole.URI
+
+	intent.Annotations = cluster.Spec.Metadata.GetAnnotationsOrNil()
+	intent.Labels = naming.Merge(
+		cluster.Spec.Metadata.GetLabelsOrNil(),
+		map[string]string{
+			naming.LabelCluster: cluster.Name,
+			naming.LabelRole:    naming.RoleCrunchyBridgeClusterPostgresRole,
+			naming.LabelCrunchyBridgeClusterPostgresRole: roleName,
+		})
+
+	err := errors.WithStack(r.setControllerReference(cluster, intent))
+
+	return intent, err
+}
+
+// reconcilePostgresRoles writes the objects necessary to manage roles and their
+// passwords in PostgreSQL.
+func (r *CrunchyBridgeClusterReconciler) reconcilePostgresRoles(
+	ctx context.Context, apiKey string, cluster *v1beta1.CrunchyBridgeCluster,
+) error {
+	_, _, err := r.reconcilePostgresRoleSecrets(ctx, apiKey, cluster)
+
+	// TODO: If we ever add a PgAdmin feature to CrunchyBridgeCluster, we will
+	// want to add the role credentials to PgAdmin here
+
+	return err
+}
+
+func (r *CrunchyBridgeClusterReconciler) reconcilePostgresRoleSecrets(
+	ctx context.Context, apiKey string, cluster *v1beta1.CrunchyBridgeCluster,
+) (
+	[]*v1beta1.CrunchyBridgeClusterRoleSpec, map[string]*corev1.Secret, error,
+) {
+	log := ctrl.LoggerFrom(ctx)
+	specRoles := cluster.Spec.Roles
+
+	// Index role specifications by PostgreSQL role name and make sure that none of the
+	// secretNames are identical in the spec
+	secretNames := make(map[string]bool)
+	roleSpecs := make(map[string]*v1beta1.CrunchyBridgeClusterRoleSpec, len(specRoles))
+	for i := range specRoles {
+		if secretNames[specRoles[i].SecretName] {
+			// Duplicate secretName found, return early with error
+			err := errors.New("There are duplicate Role SecretNames in the spec. SecretNames must be unique.")
```

Comment:
I have a vague memory of us preferring `fmt.Errorf`, let me check on that...

Yeah, I'm seeing more of that.

Reply:
Okay, I can switch to using that.

Reply:
Although when I search the repo I see 74 results for `errors.New` and 55 results for `fmt.Errorf`...

Reply:
Probably just the preference of whoever was writing the code...